Task to generate Painless API's per context

jdconrad opened the pull request:
This adds a gradle task called generateContextDoc in the Painless module.  The task will start a cluster, issue commands against the context rest api for Painless, and generate documentation for each API per context.  Each context has a first page of classes sorted by package first and class name second, along with a page per package with each classes' constructors, methods, and fields.  A link is generated for each constructor, method, and field to a JavaDoc page when possible.

The change itself is small, but the number of lines is large due to the auto-generated changes in documentation.

Follow up: 1) Internal documentation. 2) Appropriately link each context to its API docs. 3) Add static methods/bindings.  4) Add class inheritance hierarchy.  5) Links per class/package.  6) Improve method links for declaring classes.  7) Explore reduction of total classes using hierachy.  8) Open-ended exploration of more automation for documentation per context.

Concern: With all the new links, the docs build process has slowed down some.  I really hope that I don't have to reduce the number because I think the layout is significantly improved for figuring out what's available.

jdconrad committed:
Added initial gradle files.

rjernst committed:
use clusterformation instead of testclusters

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Add gradle dependencies.

jdconrad committed:
Remove old api docs.

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Build table of contexts to apis

jdconrad committed:
Add basic page outline for all apis.

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Checkpoint

jdconrad committed:
Checkpoint.

jdconrad committed:
Add indices for each context.

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Checkpoint.

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Checkpoint.

jdconrad committed:
checkpoint

jdconrad committed:
Initial generation of Painless APIs per context

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented on docs/painless/painless-api-reference/painless-api-reference-score/index.asciidoc:
```diff
@@ -0,0 +1,472 @@
+
+
+include::painless-api-reference-score-java-lang.asciidoc[]
+include::painless-api-reference-score-java-math.asciidoc[]
```
I can't find this file in the PR. Is github's UI eating it?

jdconrad replied:
They're added now :).  Forgot to add the new files here.

nik9000 commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java:
```diff
@@ -66,11 +66,11 @@
-    private final String rtn;
-    private final int readOnly;
-    private final List<String> parameters;
+    public final String declaring;
```
Its probably worth a comment about why these are public instead. It'd be nice if we didn't need it though!

jdconrad replied:
Would you prefer these be getters instead or comments here?

nik9000 replied:
Hmmmm. I think getters would be more "normal java" to be honest. Personally, I think `public final` is fine but I think we're trying to be more "normal java".... :man_shrugging:

rjernst commented on modules/lang-painless/src/doc/java/org/elasticsearch/painless/ContextDocGenerator.java:
```diff
@@ -0,0 +1,462 @@
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+public final class ContextDocGenerator {
```
A minimal javadoc explaining what this tool does would be nice for future modifications and understanding by others

jdconrad replied:
Agreed.  Will add.

jdconrad committed:
Add generated docs files.

nik9000 commented:
` +246,126 −6,464 ` Ah hahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahaha.

Wow.

nik9000 reviewed (commented):
Ah! I see the nature of the combinatorial explosion we're dealing with here. Do you think it'd be kinder to make files with the extra APIs that each context add on top of the global context? That'd save a ton of bits, speed up the docs build, and make it super clear what part of each context is unique.

rjernst reviewed (commented):
This looks ok, but I think we should see how much work deduplication is. With this change every context gets its own copy of every whitelisted class/method, while the vast majority of these are exactly the same.

jdconrad commented:
@nik9000 and @rjernst Thanks for the initial reviews.  Agree completely on consolidating the APIs where possible.  Going to add a WIP to this review for now, and work on that for the initial check in because it's clearly a bit of an increase in lines ;)

jdconrad commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:
```diff
@@ -67,7 +67,8 @@ class PluginBuildPlugin extends BuildPlugin {
 
-            if (project.plugins.hasPlugin(TestClustersPlugin.class) == false) {
+            if (project.plugins.hasPlugin(TestClustersPlugin.class) == false
+                    || project.testClusters.findByName('integTest') == null) {
```
Understood.  I'm confident @rjernst can take a look at those :)

mark-vieira replied:
What's the reason for this specific change?

jdconrad replied:
Removed as it's not needed.  Good catch.  Believe this was left behind when figuring out why the task wasn't working.

mark-vieira commented on modules/lang-painless/build.gradle:
```diff
@@ -49,17 +52,41 @@ task apiJavadoc(type: Javadoc) {
+  docCompile project(':modules:lang-painless')
+}
+
+ClusterConfiguration clusterConfig = project.extensions.create("generateContextCluster", ClusterConfiguration.class, project)
```
@atorok we should probably avoid using `ClusterFormationTasks` and instead use the `TestClustersPlugin` here if that's possible. Would that work here?

rjernst replied:
We tried using testclusters, but it there are issues with it here. I don't think it is ready for general use yet.

atorok replied:
We are using it for all plugins  and modules now, so i'm +1 on adding new tests with testclusers. Wouldn't determinedly be interested in hearing why it didn't work.

mark-vieira replied:
I agree with @atorok we are full swing migrating at this point. We should have very good reason for _adding_ new usages of `ClusterFormationTasks`. The only valid reason I could see is if test clusters does not yet support a specific capability required by these tests.

rjernst replied:
IIRC, testclusters didn't work before because painless is a module, and even though we were not using it for integTest, the check in PluginBuildPlugin would throw an error since the testclusters plugin was applied in a module. Since it is now used by modules, we may be able to get it working.

mark-vieira replied:
Yeah. I think that is supported now. @atorok can confirm.

rjernst replied:
Testclusters still does not work here. I think there needs to be more buildSrc tests for standalone uses, as I see a lot of initialization and wiring done inside PluginBuildPlugin. The configuration we tried was:

```
testClusters {
  docExtractor {
    version = project.version
    javaHome = project.file(project.ext.runtimeJavaHome)
    distribution = "DEFAULT"
  }
}

task generateContextDoc(type: JavaExec) {
  main = 'org.elasticsearch.painless.ContextDocGenerator'
  classpath = sourceSets.doc.runtimeClasspath
  useCluster testClusters.docExtractor
  systemProperty "cluster.uri", "${-> testClusters.docsExtractor.httpSocketURI}"
}
```

In addition to the extra settings of version and java home that should be defaulted to reasonable values and not require setting explicitly,  it seems to fail to depend on actually building the distribution:
> Caused by: org.elasticsearch.gradle.testclusters.TestClustersException: Can not start node{:modules:lang-painless:docExtractor-1}, missing: /mnt/ramdisk/elasticsearch/build/testclusters/extract/org.elasticsearch.distribution.elasticsearch/elasticsearch-8.0.0-SNAPSHOT

I don't think this PR should be blocked on making test clusters work for non integ tests.

mark-vieira commented on modules/lang-painless/build.gradle:
```diff
@@ -49,17 +52,41 @@ task apiJavadoc(type: Javadoc) {
+}
+clusterConfig.distribution = 'default'
+
+task generateContextDoc(type: JavaExec) {
```
If we could add proper input/outputs to this task so we could benefit from UP-TO-DATE checking and eventually the build cache that would be great.

https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_runtime_api

jdconrad replied:
I would prefer to defer this to a different PR for now.  This task is run separately from the standard build and is normally not be run in general. The infrastructure to effectively check the inputs (new script contexts, changes to resource file whitelists, and changes to the code generation) isn't fully supported, yet.  Once this task is integrated as part of the standard build, I would be happy to re-address this.

mark-vieira commented on modules/lang-painless/build.gradle:
```diff
@@ -49,17 +52,41 @@ task apiJavadoc(type: Javadoc) {
+
+sourceSets {
+  doc {
+    java {
```
This section is redundant actually. The default convention is for source to live in `src/<<source set name>>/java`.

jdconrad replied:
Fixed.

jdconrad committed:
Clean up gradle.

jdconrad commented on docs/painless/painless-contexts-api-reference.asciidoc:
```diff
@@ -0,0 +1,10 @@
+[[painless-api-reference]]
```
@nik9000 If i leave it as painless-api-reference (need to change the name of the asciidoc) will this negate the need for a redirect as saved browser links will go to the new page at the current url?

nik9000 replied:
It will. I'm fine with that. Should the file's name stay the same?

jdconrad replied:
Changed it back.  Good catch, ty!

jdconrad committed:
Checkpoint

jdconrad committed:
Checkpoint.

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Merge branch 'master' into docgen

jdconrad committed:
Create a shared API to reduce duplication across contexts.

jdconrad committed:
Add generated docs.

jdconrad committed:
Add javadoc comment.

jdconrad committed:
Add getters.

jdconrad commented:
@nik9000 and @rjernst Thanks for the reviews.  I believe I have addressed all the feedback appropriately and this is ready for another round.

@mark-vieira Are you okay with the gradle task as it is for now based on the feedback from @rjernst ?  As soon as the compatibility issues are handled, I'm happy to switch the task to use test clusters.

mark-vieira commented:
> @mark-vieira Thanks for the review. Are you okay with the gradle task as it is for now based on the feedback from @rjernst ?

That's fine. We can revisit this as part of the ongoing effort to migrate to test clusters.

jdconrad committed:
Move javadoc to class instead of method.

nik9000 commented on docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc:
```diff
@@ -0,0 +1,118 @@
+=== Aggs API for package org.elasticsearch.xpack.sql.expression.function.scalar.whitelist
+See the <<painless-api-reference-aggs, Aggs API>> for a high-level overview of all packages.
+
+[[painless-api-reference-aggs-InternalSqlScriptUtils]]
```
I think we should cut this out of the output. It isn't really "public".

jdconrad replied:
I agree with this sentiment.  I don't have a great idea for how to do this and keep it automated.  Any suggestions?

jdconrad replied:
For now I've added a hardcoded list that can be added to easily in the meantime.  In the future I would like to make this part of the entire whitelisting process where we can mark a class as internal and have it included in the PainlessLookup.  Possibly, we can even ensure that a user can't execute the method with an internal versus external compilation flag.

nik9000 replied:
:+1:

nik9000 commented on docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc:
```diff
@@ -0,0 +1,118 @@
+
+[[painless-api-reference-field-IntervalDayTime]]
+==== IntervalDayTime
+* boolean {java11-javadoc}/java.base/java/lang/Object.html#equals(java.lang.Object)[equals](Object)
```
Should there be more stuff here?

jdconrad replied:
This is a direct translation from the existing whitelist.  I do want to make it more sophisticated so that it can recognize things that may not need to be included, but that I think that's probably a different PR.

nik9000 replied:
:+1:

nik9000 commented on docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc:
```diff
@@ -0,0 +1 @@
+
```
I wonder if we shouldn't emit empty files. It isn't a big deal but I figure it'd be nice.

jdconrad replied:
Agreed, and I've thought of this issue too.  I would prefer to add this in a follow up PR as this one is already getting long.

nik9000 replied:
:+1:

jdconrad replied:
Decided to fit this one in and fix it.

nik9000 commented on docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc:
```diff
@@ -0,0 +1,9 @@
+[[painless-api-reference-interval]]
```
I think we should have a note in any generated files. I havne't checked if this is generated though. If it isn't generated, I think it probably should be.

jdconrad replied:
It's generated.  Will add a note.

nik9000 replied:
:+1:

jdconrad committed:
Removed empty specialized apis.  Added list for internal class
exclusions.

nik9000 reviewed (approved):
LGTM

I noticed in the "score" example it said `The standard Painless API is available.` and had a link to the root of the API page. This should probably mention that there is a special score API and maybe link directly to the standard API and the score API rather than the root, but I think that is a "for later" thing.

LGTM

jdconrad commented:
@nik9000 Thanks for the multiple rounds of reviews! You're right about the need for changes to the contexts documentation at this point.  I'll follow up in another PR with changes for those.

@mark-vieira Are you okay with this code being committed as is?  I didn't want to dismiss your review requesting changes without asking.
